🚀 2단계 - 리팩터링(메뉴)

hoon25 opened the pull request:
안녕하세요 리뷰어님! 2단계 미션 늦었지만 잘 부탁드립니다 🙇

## 진행과정 회고
DomainService를 `언제`, `어느 바운디드 컨텍스트 관점`으로 사용할지가 제일 고민이 많이됐네요.

### 언제 사용할까?
#### 바운디드 컨텍스트 1개
- 이전에 `ProductNamePolicy`리팩토링 리뷰를 반영했었는데, 동일하게 `ProductPricePolicy`도 추출해보니, 불필요하게 코드가 많아지는 느낌을 받아서 `ProductPricePolicy`는 다시 제거했어요.
- 리뷰어님도 미션 특성상 제안 주신거라고 느껴지긴했는데, **한 개의 바운디드 컨텍스트에서는 로직이 많이 복잡해 지지 않는 이상**, VO 추출로도 충분하지 않나 생각했어요!

#### 바운디드 컨텍스트 다수
- id로만 참조를 하게되니 자연스럽게 사용하게 되네요👍

### 어느 바운디드 컨텍스트 관점으로 사용할까?
#### 상품 가격 변경 시 메뉴 display상태 변화
- Product 관점
  - ProductPrice를 변경될 때, 해당 Product를 사용하는 Menu가 MenuPrice>MenuAmount이면 NotDisplayed한다. 
- Menu 관점
  - Menu에 속한 MenuProduct의 ProductPrice가 변경될 때, MenuPrice>MenuAmount이면 NotDisplayed한다.

**Menu관점의 모델링이 더 자연스럽다고 생각**하고 `MenuProductPriceChanged`도메인 서비스를 Menu 패키지에 뒀어요.
productId만 인자로 넘겨주니 현재는 트랜잭션에 묶여있지만, Product 상태변경에 대한 이벤트를 처리하는 방식으로 느껴지기도 했네요!

리뷰어님의 생각은 어떠신가요 ❓

hoon25 committed:
docs(Menu): Menu관련 README 추가

hoon25 committed:
refactor(Product): ProductPrice VO 적용

hoon25 committed:
feat(MenuGroup): 생성, 전체조회 추가

hoon25 committed:
refactor(MenuProduct): menuProducts 일급컬렉션 적용 및 amount 계산 로직 이동

hoon25 committed:
refactor(MenuProduct): menuProduct 관련 로직 도메인으로 이동

hoon25 committed:
refactor(Menu): MenuName VO 추출

hoon25 committed:
refactor(Menu): MenuPrice VO 추출

hoon25 committed:
feat(Menu): MenuPrice, MenuDisplay 상태변경 domain 구현

hoon25 committed:
feat(Menu): Menu생성 application 로직 구현

hoon25 committed:
refactor(MenuProduct): Product 의존성 제거 및 id참조로 변경

hoon25 committed:
test(MenuProduct): Product 의존성 제거 테스트 반영

hoon25 committed:
feat(Menu): 메뉴 등록, 가격변경
- 메뉴 등록 테스트 추가
- 메뉴가격 변경 추가

hoon25 committed:
feat(Menu): 메뉴 노출, 비노출 추가

hoon25 committed:
remove(ProductPrice): 불필요한 내부 로직 및 테스트 제거
- Comparable, ProductPricePolicy

hoon25 committed:
refactor(Menu): ProductPrice 변경 시 도메인서비스 MenuProductPriceChanged 추가

hoon25 commented on src/main/kotlin/kitchenpos/menu/tobe/domain/MenuProductPriceChanged.kt:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.*
+
+/**
```
도메인서비스의 바운디드컨텍스트 관점 고민 부분입니다!

추가적으로 본문에는 모델링이 더 자연스러워서 Menu를 선택했다고 했는데, 근거가 부족하다고 생각해서요. 더 좋은 근거를 어떻게 설명할 수 있을까요?

Gomding replied:
> 상품 가격 변경 시 메뉴 display상태 변화
> 
> Product 관점
>  - ProductPrice를 변경될 때, 해당 Product를 사용하는 Menu가 MenuPrice>MenuAmount이면 NotDisplayed한다.
> 
> Menu 관점
> - Menu에 속한 MenuProduct의 ProductPrice가 변경될 때, MenuPrice>MenuAmount이면 NotDisplayed한다.
> 
> Menu관점의 모델링이 더 자연스럽다고 생각하고 MenuProductPriceChanged도메인 서비스를 Menu 패키지에 뒀어요.
> productId만 인자로 넘겨주니 현재는 트랜잭션에 묶여있지만, Product 상태변경에 대한 이벤트를 처리하는 방식으로 느껴지기도 했네요!

상품의 가격이 변경됐을 때 메뉴의 노출 여부를 결정하는것은 
`Menu 의 정책` VS `Product 의 정책` 관점에서 비교해보는 방법이 있을것 같아요 :)

또 Product 관점에서 보면 Menu 의 존재를 몰라도 된다고 볼 수 있죠 
Product 는 Menu 라는 개념 없이도 존재할 수 있으니까요 :)

정리하면 누구의 정책인가 관점에서 본다면 좋을 것 같아요!

Gomding commented on src/main/kotlin/kitchenpos/product/tobe/domain/ProductPrice.kt:
```diff
@@ -0,0 +1,30 @@
+import java.math.BigDecimal
+
+@Embeddable
+class ProductPrice(
```
> 이전에 ProductNamePolicy리팩토링 리뷰를 반영했었는데, 동일하게 ProductPricePolicy도 추출해보니, 불필요하게 코드가 많아지는 느낌을 받아서 ProductPricePolicy는 다시 제거했어요.
> 리뷰어님도 미션 특성상 제안 주신거라고 느껴지긴했는데, 한 개의 바운디드 컨텍스트에서는 로직이 많이 복잡해 지지 않는 이상, VO 추출로도 충분하지 않나 생각했어요!

추출해보고 삭제하는 과정에서 창훈님의 기준을 만드신것 같아서
결과적으로 좋았다고 생각이 드네요 :)
저도 VO 로 만드는것으로 충분하다고 생각합니다!

Gomding commented on src/main/kotlin/kitchenpos/product/tobe/application/ProductService.kt:
```diff
@@ -35,16 +37,9 @@ class ProductService(
-                }
-            }
-        }
+        product.changePrice(ProductPrice(request.price))
+
+        menuProductPriceChanged.changedProduct(productId)
```
만약 메뉴 노출여부를 수정하다가 예외가 발생했다면
상품의 가격도 함께 복구해줘야할 것 같아요! 😃 

ACID가 잘 지켜졌는지 확인해보죠!

hoon25 replied:
`ProductServiceIntegrationTest`에 추가해뒀어요!

Gomding commented on src/main/kotlin/kitchenpos/product/tobe/application/ProductService.kt:
```diff
@@ -2,27 +2,29 @@ package kitchenpos.product.tobe.application
     private val productRepository: ProductRepository,
-    private val menuRepository: MenuRepository,
     private val profanities: Profanities,
+    private val menuProductPriceChanged: MenuProductPriceChanged,
```
Product 는 Menu 에 대한 존재를 몰라도 되므로
`ProductPriceChangedListener` 라는 이름이 더 적절할 수도 있을 것 같아요 :)

또는 `도메인 이벤트` 를 사용하는 방식도 고려해볼 수 있어요 :)

hoon25 replied:
도메인이벤트 방식으로 적용했어요!

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/infra/DefaultMenuAmountService.kt:
```diff
@@ -0,0 +1,24 @@
+) : MenuAmountService {
+
+
+    override fun amount(menuProducts: MenuProducts): BigDecimal {
+        var sum = BigDecimal.ZERO
+        menuProducts.menuProducts.forEach { menuProduct ->
+            val product = productRepository.findById(menuProduct.productId!!)
+                .orElseThrow { NoSuchElementException("상품을 찾을 수 없습니다.") }
+            sum += product.productPrice.price * menuProduct.quantity.toBigDecimal()
+        }
+        return sum
+    }
```
상품에 대한 의존이 필요하여 구현해주신것으로 보이네요 👍 👍 👍  💯 

1. `Product` 와 소통하는것이 목표라면 `ProductClient` 라는 이름도 있었을 것 같아요 :)
2. 상품의 가격만 가져오는 방법도 있었을 것 같은데 합을 구하도록 구현하신 부분도 좋은것 같아요!
어떤 관점에서 구현하셨는지 궁금해요~

hoon25 replied:
> 1. `Product` 와 소통하는것이 목표라면 `ProductClient` 라는 이름도 있었을 것 같아요 :)

Menu관점으로 `ProductClient`로 재정의해서 사용한다고 생각하고 적용했는데 맞을까요?

hoon25 replied:
> 2. 상품의 가격만 가져오는 방법도 있었을 것 같은데 합을 구하도록 구현하신 부분도 좋은것 같아요!
>    어떤 관점에서 구현하셨는지 궁금해요~

첫 관점) 처음에는 Menu에서 Amount계산이 자주 사용되니, **반복되는 로직을 묶고**, **유닛테스트에 mocking도 용이**하게 하자 라는 관점

바뀐 관점) 리뷰반영 하면서 좀 더 고민 해보니, **Domain에 들어가는 DomainService들이 좀 더 Domain의 느낌이 묻어나면 좋겠구나**라는 관점으로 바뀜

바뀐관점을 기준으로 제가 생각한 장점을 적어보자면

- ex) Profanities 대신 ***NamePolicy를 사용하는 것
  - `Profanities`가 `NamePolicy`에서만 사용된다고 직관적으로 알게됨
  - `Profanities`의 변경에 Domain이 영향을 받지 않게 됨

- ex) ProductClient 대신 MenuAmountService
  - `fun display(productClient: ProductClient)` -> `fun display(menuAmountService: MenuAmountService)`
  - 메뉴를 보여줄 때, 전자는 **Product정보가 필요하구나**에 그치지만, 후자는 **Menu의 금액이 필요하구나**라고 좀 더 직관적으로 알게됨


결국 Profanities, ProductClient 같은 상대적으로 상위 개념들을 **도메인스럽게 한번 더 묶는것**이 좋겠구나 생각하게 되었네요.

Gomding replied:
> 메뉴를 보여줄 때, 전자는 Product정보가 필요하구나에 그치지만, 후자는 Menu의 금액이 필요하구나라고 좀 더 직관적으로 알게됨

반대로 메뉴를 보여줄 때 메뉴의 금액이 필요하다는 정보는 추측이 가능할 것 같은데
`상품의 정보가 필요하구나` 라는 정보가 더 추측하기 어렵다고 생각했어요! 🤔 
그래서 ProductClient 로 표현하는게 더 좋겠다는 의견입니다 ㅎㅎ

+) `MenuProductAmountService` -> `ProductClient` depth 가 더 들어가서 파악이 힘든것은 없을까요!
그리고 외부와 소통할때마다 Service 가 하나씩 늘어날 것 같아요!
요런 관점은 어떻게 생각하시나요?

hoon25 replied:
> `MenuProductAmountService` -> `ProductClient` depth 가 더 들어가서 파악이 힘든것은 없을까요! 그리고 외부와 소통할때마다 Service 가 하나씩 늘어날 것 같아요! 요런 관점은 어떻게 생각하시나요?

저도 `MenuProductAmountService`같은 Service는 항상 만들기 보다는, **`반복사용`, `복잡한 로직`, `도메인적으로 이해도`가 높은 로직**에 대해서만 만들것 같아요.
자주 사용되는 컴포넌트에 대해서 유연성을 포기하고, 강결합하겠다는 의도였습니다

---
여기서부터는 DM으로 소통한 내용을 정리할게요!

**`ProductClient`는 `OutboundPort`로 보는게 더 적절하기에 `Domain`이 의존하는게 적절하지 않다고 생각**

**조회의 책임은 application계층**에 두고, **domain계층에서는 계산만 수행**하도록 변경
ProductClient는 Menu에서 다시한번 정의한 `ProductInfo` VO를 반환하고, Menu도메인에서는 ProductInfo를 사용하도록 함

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/domain/Menu.kt:
```diff
@@ -35,32 +38,44 @@ class Menu(
+        validateMenuPrice(menuAmountService, menuPrice)
+    }
+
+    fun amount(menuAmountService: MenuAmountService): BigDecimal {
+        return menuAmountService.amount(menuProducts)
+    }
```
사용하지 않는 메서드라 제거해도 되겠군요! 😃

hoon25 replied:
삭제했어요!

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/domain/MenuProducts.kt:
```diff
@@ -0,0 +1,23 @@
+import jakarta.persistence.OneToMany
+
+@Embeddable
+class MenuProducts(
```
일급 컬렉션 👍 👍

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/domain/MenuName.kt:
```diff
@@ -0,0 +1,29 @@
+import jakarta.persistence.Embeddable
+
+@Embeddable
+class MenuName(
```
값 객체로 꼼꼼하게 만들어주셨네요 😃

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/domain/MenuNamePolicy.kt:
```diff
@@ -0,0 +1,10 @@
+
+import kitchenpos.common.domain.Profanities
+
+class MenuNamePolicy(private val profanities: Profanities) {
```
MenuNamePolicy 와 같은 도메인 서비스도 스프링 빈으로 관리해봐도 좋을 것 같아요 :)

추가로 `@DomainService` 라는 애너테이션을 사용해봐도 좋구요!

hoon25 replied:
interface 추출하고 `@DomainService` 만들어서 적용했어요!

Gomding reviewed (changes requested):
안녕하세요 창훈님!
2단계 구현하느라 고생하셨어요 👍 👍 
깔끔하게 구현해주셨네요!
몇 가지 코멘트 남겼으니 의견 나눠보면 좋겠어요 🙏 

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요!

hoon25 committed:
refactor(Product, Menu): Product 가격 변경 Event 발행

hoon25 committed:
refactor(Menu): MenuAmount 계산 시 Menu의 ProductClient를 사용

hoon25 committed:
move(MenuProductPrice): MenuProductPrice 상태변경 로직 EventListener로 이동

hoon25 committed:
rename(Menu): ProductPriceChangedEventListener to ProductPriceChangedListener

hoon25 committed:
refactor(DomainService): Bean으로 관리되는 DomainService에 @DomainService 적용

hoon25 committed:
refactor(Menu, Product): ***NamePolicy interface 추출

hoon25 committed:
remove(Menu): unused method

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/infra/DefaultMenuProductPriceChangedListener.kt:
```diff
@@ -0,0 +1,27 @@
+
+    @TransactionalEventListener(phase = TransactionPhase.BEFORE_COMMIT)
+    override fun handle(event: ProductPriceChangedEvent) {
+        val menus = menuRepository.findAllByProductId(event.productId)
+        menus.forEach {
+            if (!it.canDisplay(amountService)) {
+                it.notDisplay()
+            }
+        }
```
변경된 상품의 가격에 맞춰 메뉴의 가격도 수정이 필요하지 않을까요! 😃

hoon25 replied:
`상품가격`이 변경되었을때, `메뉴가격>메뉴금액`인 경우 `메뉴`를 `비노출` 상태로 변경한다.

메뉴의 Displayed 상태만 변경하는 것으로 알고있는데, 제가 놓친 부분이 있을까요?

Gomding replied:
앗 혼란드려 죄송합니다 제가 헷갈렸네요 😢

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/infra/DefaultMenuAmountService.kt:
```diff
@@ -0,0 +1,17 @@
+import kitchenpos.menu.tobe.domain.MenuProducts
+import kitchenpos.menu.tobe.domain.ProductClient
+
+@DomainService
+class DefaultMenuAmountService(
```
해당 객체는 infra 계층이 어울리는지 궁금해요!

+) 인터페이스를 두는것이 꼭 필요한지도 고민해볼 수 있겠어요! 🙏

hoon25 replied:
DomainService라고 반드시 interface를 둘 필요는 없을 것 같네요! 

**interface삭제**하고 domain 계층으로 구현체를 옮겼어요.

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/infra/DefaultMenuNamePolicy.kt:
```diff
@@ -0,0 +1,13 @@
+import kitchenpos.common.domain.Profanities
+import kitchenpos.menu.tobe.domain.MenuNamePolicy
+
+@DomainService
+class DefaultMenuNamePolicy(private val profanities: Profanities) : MenuNamePolicy {
```
해당 부분도 동일합니다 :)

hoon25 replied:
동일하게 interface 삭제하고 구현체를 domain계층으로 이동했어요.

Gomding reviewed (changes requested):
안녕하세요 창훈님!
피드백 잘 반영해주셨네요 👍 
추가로 의견 필요한 부분이 있어 코멘트 남겼어요
궁금한 점 있으면 언제든 DM이나 코멘트 남겨주세요~

hoon25 committed:
refactor(Menu, Product): DomainService 관련 interface 및 패키지 이동
- ***EventListener 인터페이스 삭제 및 구현체 application계층으로 이동
- ***NamePolicy 인터페이스 삭제 및 구현체 domain계층으로 이동

hoon25 committed:
refactor(Menu): MenuAmountService 사용하는 곳을 ProductClient로 변경

hoon25 committed:
remove(Menu): 사용하지 않는 MenuAmountService삭제

hoon25 committed:
refactor(Menu): Menu의 Product조회 책임 변경 (domain -> application)
- ProductClient의 return type을 MenuInfo VO로 변경
- Menu Domain계층의 ProductClient 의존성 삭제 및 MenuInfo를 사용하도록 변경

Gomding commented on src/main/kotlin/kitchenpos/menu/tobe/application/MenuService.kt:
```diff
@@ -23,7 +23,7 @@ class MenuService(
     private val productRepository: ProductRepository,
     private val menuNamePolicy: MenuNamePolicy,
-    private val menuAmountService: MenuAmountService,
+    private val productClient: ProductClient,
```
👍 👍 👍

Gomding reviewed (approved):
안녕하세요 창훈님!
메뉴 리팩터링 구현 잘해주셨어요 👍 👍 👍 
이번 단계는 이만 머지하겠습니다!
마지막 단계도 화이팅입니다 🎉
